File: 2018.8.5/Vjudge/Failed Scripts/The knapsack problem.cpp
#include<stdio.h>
#include<iostream>
using namespace std;
int S, N;//Size, Number
int s[2010], v[2010]; //size, value
int cache[2005][2005];
int solve(int S, int i)
{
	if (cache[S][i] != -1) return cache[S][i];
	if (S < 0 || i < 0) return 0;
	int got;
	if(S - s[i] >= 0)
		got = max(solve(S - s[i], i - 1) + v[i], solve(S, i - 1));
	else got = solve(S, i - 1);
	if (cache[S][i] == -1) cache[S][i] = got;
	return got;
}
void tester()
{
	scanf("%d %d", &S, &N);
	for(int i = 0; i < N; i++)
		scanf("%d %d", &s[i], &v[i]);
	int Max = solve(S, N - 1);
	printf("%d\n", Max);
}

int main()
{
	for(int i = 0; i < 2000; i++)
		for (int j = 0; j < 2000; j++)
			cache[i][j] = -1;
	tester();
	return 0;
}
	

File: 2018.8.5/Vjudge/Failed Scripts/The knapsack problem.py
import sys
def tester():
    S, N = list(map(int, input().split()))
    s, v =[], []
    for i in range(N):
        get = list(map(int, input().split()))
        s.append(get[0]), v.append(get[1])
    Max = solve(S, s, v, i = N - 1)
    print(Max)

cache = {}
def solve(S, s, v, i):
    if (S, i) in cache:
        return cache[(S, i)]
    if S < 0 or i < 0: return 0
    if S - s[i] >= 0:
        get = max(solve(S - s[i], s, v, i - 1) + v[i], solve(S, s, v, i-1))
    else:
        get = solve(S, s, v, i - 1)
    if (S, i) not in cache: cache[(S, i)] = get
    return get


tester()
